### PR TITLE
Windows support via kubeadm

### LINKED ISSUE #153: Support for windows worker nodes
/kind feature

Since kubernetes 1.14 just GA windows worker nodes, add support for the same in cluster api

**Describe the solution you'd like**
1. VM Extension service needs to be refactored to handle CSE in Microsoft.Compute namespace https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/f53efdb852dab776e412e4c0cd5a91cc3f760019/pkg/cloud/azure/services/virtualmachineextensions/virtualmachineextensions.go#L69
2. Change default CNI to use Flannel, which has windows support (part of addons)
2. Change VM Startup scripts to include support for windows, following guide here https://github.com/Azure/aks-engine/blob/1242dcc6fa5f7a55c685eb94fd82d8f668e90965/parts/k8s/kuberneteswindowssetup.ps1

**Anything else you would like to add:**
aks-engine has successful implementation, following the process would help implementing it here.

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #153 

Implements https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200804-windows-support.md

**Special notes for your reviewer**:
Uses an image from https://github.com/kubernetes-sigs/image-builder/pull/382. This PR is close to merging and then this can go in.

Similar to #1035 

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Add windows support 
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
is MachinePool VMSS support non-goal/future work?

**Comment by jsturtevant** [comment]
I took a look and VMSS was pretty straight forward to add support.

File: cloud/defaults.go
Comment (jsturtevant) [inline cloud/defaults.go:214]: The Marketplace Gallery Image is being published. As soon as it is I will update this section with the Market Place info

**Comment by jsturtevant** [comment]
@CecileRobertMichon @marosset @ksubrmnn @nader-ziada This is ready for review.  The tests should run with the SIG while I wait for the Market Place images to finish publishing. That should be a minor change in the default image so reviews can get started.

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:1]: make sure you also add this page to SUMMARY.md so it shows up in the book

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:93]: how hard would it be to make this work with the kubectl plugin you wrote? Maybe open an issue as help wanted for someone to pick up?

File: test/e2e/azure_test.go
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:403]: does this need the linux machine pool? If so would it make sense to merge the test with https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1036/files#diff-93ce65d81f5c1510135f6035d43f678bf9b5962b8e4ee45105e3345a1852efc6L261 (I know that one has 1 control plane right now but we could change it to 3).

File: test/e2e/azure_test.go
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:354]: Same question about merging tests here. Are we adding more test coverage by testing a cluster _without_ Windows nodes?

File: docs/book/src/topics/windows.md
Comment (jsturtevant) [reply docs/book/src/topics/windows.md:93]: It already works with the plugin.  This is just manual steps for users that might not have the plugin, I could add a note about the plugin here?

File: test/e2e/azure_test.go
Comment (jsturtevant) [reply test/e2e/azure_test.go:403]: I would prefer to use the existing cluster tests, but since Windows is using Flannel it needed it's own cluster.  Calico support for Windows became available in Windows November updates and is going to require some effort to get it to work with kubeadm.

**Comment by jsturtevant** [comment]
Tests failed due to the SIG Image not being available in the random region picked by the tests. I am updating the SIG to replicate to the testing regions.  Market Place image is still provisioning.

File: cloud/defaults.go
Comment (jsturtevant) [reply cloud/defaults.go]: The image is now published: `az vm image list --publisher cncf-upstream --offer capi-windows -o table --all`

File: docs/book/src/topics/windows.md
Comment (jsturtevant) [reply docs/book/src/topics/windows.md:93]: add a note to the ssh section

File: templates/flavors/windows/kustomization.yaml
Comment (cpanato) [inline templates/flavors/windows/kustomization.yaml:5]: nit: new line

**Comment by jsturtevant** [comment]
The Market place images are now published for the e2e test.  I will also begin the process for the new versions released yesterday.

```
az vm image list --publisher cncf-upstream --offer capi-windows -o table --all

Offer         Publisher      Sku                          Urn                                                                Version
------------  -------------  ---------------------------  -----------------------------------------------------------------  ----------
capi-windows  cncf-upstream  k8s-1dot19dot3-windows-2019  cncf-upstream:capi-windows:k8s-1dot19dot3-windows-2019:2020.12.07  2020.12.07
capi-windows  cncf-upstream  k8s-1dot19dot4-windows-2019  cncf-upstream:capi-windows:k8s-1dot19dot4-windows-2019:2020.12.09  2020.12.09
```

/test pull-cluster-api-provider-azure-e2e

**Comment by jsturtevant** [comment]
Good news is the VM's got deployed with the Market place images.  I've run the tests a few times locally (`scripts/ci-e2e.sh`) and didn't see these errors.  It might help once https://github.com/kubernetes/test-infra/pull/20197 

/test pull-cluster-api-provider-azure-e2e

**Comment by jsturtevant** [comment]
All the Windows tests passed! It was the ipv6 that failed.  Splitting out the jobs as planned will help.

```
 [Fail] Workload cluster creation Creating a ipv6 control-plane cluster [It] With ipv6 worker node 
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/helpers.go:196

Ran 6 of 16 Specs in 5950.430 seconds
FAIL! -- 5 Passed | 1 Failed | 0 Pending | 10 Skipped 
```

**Comment by cpanato** [comment]
looks like good, if im not crazy

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:9]:
```suggestion
## Deploy a workload
```

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:82]: maybe this should point to the custom images doc for a reference on building your own images

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:106]:
```suggestion
### Kube-proxy and CNIs
```

File: docs/book/src/topics/windows.md
Comment (CecileRobertMichon) [inline docs/book/src/topics/windows.md:108]:
```suggestion
Kube-proxy and Windows CNIs are deployed via Cluster Resource Sets.  Windows doesn't not have a kube-proxy image due 
```

File: cloud/services/scalesets/scalesets_test.go
Comment (CecileRobertMichon) [inline cloud/services/scalesets/scalesets_test.go:381]: heads up @devigned this conflicts with #1067

File: cloud/scope/machinepool.go
Comment (CecileRobertMichon) [inline cloud/scope/machinepool.go:127]: is this a VMSS restriction or a Windows hostname restriction? asking because I noticed for Machines it's 15 chars

File: cloud/scope/machine.go
Comment (CecileRobertMichon) [inline cloud/scope/machine.go:265]: Let's document this in the book

**Comment by CecileRobertMichon** [comment]
So I used tilt to deploy two windows clusters, 1 with machine pools and one without, and the VMs got created fine and nodes joined the cluster but the windows nodes are NotReady:

```
k --kubeconfig ./kubeconfig get nodes
NAME                                           STATUS     ROLES    AGE     VERSION
machinepool-win-template-control-plane-zj8b8   Ready      master   5m14s   v1.19.4
machinepool-win-template-mp-0000002            Ready      <none>   2m34s   v1.19.4
machinepool-win-template-mp-0000003            Ready      <none>   2m47s   v1.19.4
winp-win000000                                 NotReady   <none>   2m      v1.19.4
winp-win000003                                 NotReady   <none>   115s    v1.19.4
```

```
k --kubeconfig ./kubeconfig get nodes
NAME                               STATUS     ROLES    AGE     VERSION
win-templ-frc6p                    NotReady   <none>   3m17s   v1.19.4
win-templ-zbpfw                    NotReady   <none>   3m16s   v1.19.4
win-template-control-plane-2jhng   Ready      master   6m38s   v1.19.4
win-template-md-0-6lz4c            Ready      <none>   4m35s   v1.19.4
win-template-md-0-tfp4m            Ready      <none>   4m27s   v1.19.4
```

File: docs/book/src/topics/windows.md
Comment (jsturtevant) [reply docs/book/src/topics/windows.md:82]: Will do, I think this was left over before we merged the PR.

File: cloud/scope/machinepool.go
Comment (jsturtevant) [reply cloud/scope/machinepool.go:127]: VMSS only, This is becuase VMSS adds unique value to the end of each instance

File: cloud/services/scalesets/scalesets_test.go
Comment (jsturtevant) [reply cloud/services/scalesets/scalesets_test.go:381]: Is there anything I need to do here?

File: test/e2e/azure_test.go
Comment (jsturtevant) [reply test/e2e/azure_test.go:354]: This is due to flannel being the currently supported cni.  We have split the test jobs for now and as calico support comes online will re-address this

**Comment by jsturtevant** [comment]
> windows nodes are NotReady

This is due to https://github.com/kubernetes-sigs/cluster-api/issues/4016.  I have applied a fix for the CI but missed pushing the fix for tilt.  I will update based on the comments above and include this fix.

**Comment by jsturtevant** [comment]
The e2e job failed for different set of tests each time

/test pull-cluster-api-provider-azure-e2e

**Comment by jsturtevant** [comment]
> The e2e job failed for different set of tests each time

created #1095 for further investigation

**Comment by jsturtevant** [comment]
looks like there was a flake  grabbing the kubectl version

```
mkdir -p /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin
rm -f "/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/kubectl-v1.16.13*"
curl -fsL https://storage.googleapis.com/kubernetes-release/release/v1.16.13/bin/linux/amd64/kubectl -o /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/kubectl-v1.16.13
make[1]: *** [Makefile:250: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/kubectl-v1.16.13] Error 56
```

I don't believe that is the correct version that is needed since our tests run 1.19.4.  I will follow up with an issue/pr to fix that.

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by devigned** [comment]
This PR looks great. I'd love to get it pulled in sooner than later, but I think we should hold off until after the holiday to make sure everyone gets a final review.

Awesome work, @jsturtevant!

**Review by cpanato — APPROVED** [review]
/lgtm
thanks for this! 🎉

**Comment by CecileRobertMichon** [comment]
```
k --kubeconfig ./kubeconfig get nodes
NAME                               STATUS   ROLES    AGE     VERSION
win-templ-lndvv                    Ready    <none>   2m43s   v1.19.4
win-templ-xk5fx                    Ready    <none>   2m10s   v1.19.4
win-template-control-plane-k5wgf   Ready    master   6m      v1.19.4
win-template-md-0-456pg            Ready    <none>   3m44s   v1.19.4
win-template-md-0-xfk4g            Ready    <none>   3m49s   v1.19.4
```

Nice :) 

/lgtm
/approve
🚢

**Pull request merged**
